### PR TITLE
Fix handling of INI and ENV changes during and post runtime

### PR DESCRIPTION
**Pull request by krakjoe**
### Description

Effecting setenv/htaccess.

### Readiness checklist
- [ ] (only for Members) Changelog has been added to the release document.
- [x] Tests added for this feature/bug.

### Reviewer checklist
- [ ] Appropriate labels assigned.
- [ ] Milestone is set.
- [ ] Changelog has been added to the release document. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

File: zend_abstract_interface/config/php5/config_runtime.c [hunk 1]
```diff
@@ -21,21 +21,53 @@ void zai_config_replace_runtime_config(zai_config_id id, zval *value) {
...
+}
+
 void zai_config_runtime_config_ctor(void) {
     if (runtime_config_initialized == true) return;
     runtime_config = emalloc(sizeof(zval *) * ZAI_CONFIG_ENTRIES_COUNT_MAX);
 
     for (uint8_t i = 0; i < zai_config_memoized_entries_count; i++) {
-        if (Z_TYPE(zai_config_memoized_entries[i].decoded_value) == IS_ARRAY) {
-            // arrays will land in the GC root buffer and must actually be of type struct _zval_gc_info, which also
-            // must not be shared across threads, hence forcing a copy of arrays
```
Comment (bwoebi) [inline zend_abstract_interface/config/php5/config_runtime.c:31]: Is there a reason for dropping this comment while moving it into the other function?

File: zend_abstract_interface/config/php7-8/config_runtime.c [hunk 2]
```diff
@@ -20,12 +20,48 @@ void zai_config_replace_runtime_config(zai_config_id id, zval *value) {
     ZVAL_COPY(rt_value, value);
 }
 
+static inline zval *zai_config_runtime_config_value(zai_config_memoized_entry *memoized, zval *updated) {
+    ZAI_ENV_BUFFER_INIT(buf, ZAI_ENV_MAX_BUFSIZ);
+
+    int16_t name_index = 0;
+    for (; name_index < memoized->names_count; name_index++) {
```
Comment (bwoebi) [inline zend_abstract_interface/config/php7-8/config_runtime.c:27]: name_index is not used after the loop, could you put it into the for stmt?

File: zend_abstract_interface/config/php5/config_runtime.c (hunk 1 above)
Comment (krakjoe) [reply zend_abstract_interface/config/php5/config_runtime.c:31]: no actually, just over-zealousness ...

File: zend_abstract_interface/config/php7-8/config_runtime.c (hunk 2 above)
Comment (krakjoe) [reply zend_abstract_interface/config/php7-8/config_runtime.c:27]: I must have been searching the wrong file, I found a use outside the loop the first time I looked 😕 

moving it ...

**Review by bwoebi — APPROVED** [review]
Looks good now, I like the minimally invasive way it has been implemented :-)

File: zend_abstract_interface/config/php7-8/config_runtime.c [hunk 3]
```diff
@@ -20,12 +20,47 @@ void zai_config_replace_runtime_config(zai_config_id id, zval *value) {
     ZVAL_COPY(rt_value, value);
 }
 
+static inline zval *zai_config_runtime_config_value(zai_config_memoized_entry *memoized, zval *updated) {
+    ZAI_ENV_BUFFER_INIT(buf, ZAI_ENV_MAX_BUFSIZ);
+
+    for (int16_t name_index = 0; name_index < memoized->names_count; name_index++) {
```
Comment (bwoebi) [inline zend_abstract_interface/config/php7-8/config_runtime.c:26]: We must skip this loop for system inis. System inis are recognized via `memoized->ini_change == zai_config_system_ini_change`.

File: ext/php7/configuration.c [hunk 4]
```diff
@@ -119,9 +115,6 @@ void ddtrace_config_first_rinit() {
     zend_string *internal_functions_old = zend_string_copy(
         internal_functions_ini->modified ? internal_functions_ini->orig_value : internal_functions_ini->value);
 
-    zai_config_first_time_rinit();
-    zai_config_rinit();
-
```
Comment (bwoebi) [inline ext/php7/configuration.c:124]: I believe you can just clean that whole function out then

File: zend_abstract_interface/config/php5/config_runtime.c [hunk 5]
```diff
@@ -56,10 +123,8 @@ zval *zai_config_get_value(zai_config_id id) {
         TSRMLS_FETCH();
         return &EG(uninitialized_zval);
     }
-    if (runtime_config[id] == NULL) {
-        assert(false && "runtime config is not yet initialized");
```
Comment (bwoebi) [inline zend_abstract_interface/config/php5/config_runtime.c:60]: Is there a particular reason for dropping this assertion? If *runtime* config is accessed outside of a request stage that sounds like a violation of the config model

File: zend_abstract_interface/config/php7-8/config_runtime.c [hunk 6]
```diff
@@ -10,44 +10,110 @@
...
+        if (!(GC_FLAGS(Z_COUNTED_P(zv)) & IS_STR_PERSISTENT)) {
+            zval_ptr_dtor(zv);
+        }
+    }
+}
 
 void zai_config_replace_runtime_config(zai_config_id id, zval *value) {
     zval *rt_value = &runtime_config[id];
-    zval_ptr_dtor(rt_value);
+    zai_config_runtime_dtor(rt_value);
```
Comment (bwoebi) [inline zend_abstract_interface/config/php7-8/config_runtime.c:42]: This doesn't sound very good if this needs specialized destruction. Especially if the value is leaked within dd_trace_env_config() into userland and then replaced thanks to an ini set, userland will have that zval and WILL call zval_ptr_dtor on it.

**Pull request closed without merging**